src/garage/torch/distributions/tanh_normal.py

```diff
@@ -0,0 +1,290 @@
+"""A Gaussian distribution with tanh transformation."""
+from collections import namedtuple
+
+import torch
+from torch.distributions import Normal
+from torch.distributions.independent import Independent
+
+
+class TanhNormal(torch.distributions.Distribution):
+    """A gaussian distribution that has a tanh transformation applied to it.
+
+    Algorithms like SAC and Pearl use this transformed distribution.
+    It can be thought of as a distribution of X where
+        Y ~ N(mean, cov)
+        X ~ tanh(Y)
+
+    Args:
+        mean (torch.tensor): The mean of this distribution.
+        std (torch.tensor): The stdev of this distribution.
+
+    """
+
+    def __init__(self, mean, std):
+        self._normal = Independent(Normal(mean, std), 1)
+        super().__init__()
+
+    def log_prob(self, value, pre_tanh_value=None, epsilon=1e-6):
+        """The log likelihood of a sample on the this Tanh Distribution.
+
+        Args:
+            value (torch.Tensor): The sample whose loglikelihood is being
+                computed.
+            pre_tanh_value (torch.Tensor): The value prior to having the tanh
+                function applied to it but after it has been sampled from the
+                normal distribution.
+            epsilon (float): Stabilization coefficient.
+
+        Notes:
+            - when pre_tanh_value is None, an estimate is made of what the
+              value is. This leads to a worse estimation of the log_prob.
+              If the value being used is collected from functions like
+              `sample` and `rsample`, one can instead use functions like
+              `sample_return_pre_tanh_value` or
+              `rsample_return_pre_tanh_value`
+
+
+        Returns:
+            torch.Tensor: The log likelihood of value on the distribution.
+
+        """
+        # pylint: disable=arguments-differ
+        if pre_tanh_value is None:
+            pre_tanh_value = torch.log((1 + value) / (1 - value)) / 2
+        norm_lp = self._normal.log_prob(pre_tanh_value)
+        ret = (norm_lp - torch.sum(
+            torch.log(self._clip_but_pass_gradient((1. - value**2)) + epsilon),
+            axis=-1))
+        return ret
+
+    def sample_return_pre_tanh_value(self, sample_shape=torch.Size()):
+        """Return a sample, sampled from this Tanh Normal Distribution.
+
+        Returns the sampled value before the Tanh transform is applied and the
+        sampled value with the Tanh transform applied to it.
+
+        Args:
+            sample_shape (list): Shape of the returned value.
+
+        Note: Gradients `do not` pass through this operation.
+
+        Returns:
+            action_infos: named tuple with fields
```

Comment:
returning a namedtuple from an ordinary method with multiple return values is not very pythonic. this just adds overhead. just use multiple returns.

returning a tuple in Python is equivalent to multiple returns, so you should document this as-such

```python
"""
    Returns:
        torch.Tensor: Samples from this distribution.
        torch.Tensor: Samples from the underlying :obj:`torch.distributions.Normal` distribution, prior to being transformed with `tanh`.